concordium-consensus/tests/globalstate/GlobalStateTests/BlobStore.hs

```diff
@@ -0,0 +1,258 @@
+    genDone = return Done
+    genStore = do
+        bs <- BS.pack <$> arbitrary
+        cont <- genBlobStoreOperation' (BS.length bs Seq.<| stored) cbks
+        return $ Store bs cont
+    genLoad = do
+        l <- chooseInt (0, Seq.length stored - 1)
+        cont <- genBlobStoreOperation' stored cbks
+        return $ Load l cont
+    genFlush = Flush <$> genBlobStoreOperation' stored cbks
+    genLoadBlobPtr = do
+        l <- chooseInt (0, Seq.length stored - 1)
+        let len = Seq.index stored l
+        offset <- chooseInt (0, len)
+        takeLen <- chooseInt (0, len - offset)
+        cont <- genBlobStoreOperation' stored cbks
+        return $ LoadBlobPtr l (fromIntegral offset) (fromIntegral takeLen) cont
+    genCallbacks = do
+        cont <- genBlobStoreOperation' stored (cbks + 1)
+        return $ GetCallbacks cont
+    genStoreViaCallbacks = do
+        i <- chooseInt (0, cbks - 1)
+        bs <- BS.pack <$> arbitrary
+        cont <- genBlobStoreOperation' (BS.length bs Seq.<| stored) cbks
+        return $ StoreViaCallbacks i bs cont
+    genLoadViaCallbacks = do
+        i <- chooseInt (0, cbks - 1)
+        j <- chooseInt (0, Seq.length stored - 1)
+        cont <- genBlobStoreOperation' stored cbks
+        return $ LoadViaCallbacks i j cont
+
+-- |Generator for 'BlobStoreOperation' that should ensure that invariants are established.
+genBlobStoreOperation :: Gen BlobStoreOperation
+genBlobStoreOperation = genBlobStoreOperation' Seq.empty 0
+
+instance Arbitrary BlobStoreOperation where
+    arbitrary = genBlobStoreOperation
+    shrink = shrinkBlobStoreOperation
+
+-- |Check that a 'BlobStoreOperation' satisfies basic invariants.
+bsoOKTest :: BlobStoreOperation -> Bool
+bsoOKTest = tst Seq.empty 0
+  where
+    tst _ _ Done = True
+    tst s n (Store bs cont) = tst (bs Seq.<| s) n cont
+    tst s n (Load i cont) = 0 <= i && i < Seq.length s && tst s n cont
+    tst s n (Flush cont) = tst s n cont
+    tst s n (LoadBlobPtr j off len cont) = case Seq.lookup j s of
+        Nothing -> False
+        Just bs ->
+            0 <= off
+                && 0 <= len
+                && off <= fromIntegral (BS.length bs)
+                && off + len <= fromIntegral (BS.length bs)
+                && tst s n cont
+    tst s n (GetCallbacks cont) = tst s (n + 1) cont
+    tst s n (StoreViaCallbacks i bs cont) =
+        0 <= i && i < n && tst (bs Seq.<| s) n cont
+    tst s n (LoadViaCallbacks i j cont) =
+        0 <= i && i < n && 0 <= j && j < Seq.length s && tst s n cont
+
+foreign import ccall "dynamic" callLoadCallback :: LoadCallback -> LoadCallbackType
+foreign import ccall "dynamic" callStoreCallback :: StoreCallback -> StoreCallbackType
+
+foreign import ccall "return_value_to_byte_array" vectorToByteArray :: Ptr Vec -> Ptr CSize -> IO (Ptr Word8)
+foreign import ccall unsafe "free_array_len" freeByteArray :: Ptr Word8 -> Word64 -> IO ()
+foreign import ccall unsafe "box_vec_u8_free" freeVector :: Ptr Vec -> IO ()
```

Comment:
Why do we need to re-import these here. I am not a fan of duplicating this.

Reply:
"free_array_len" we could import from `Concordium.Crypto.FFIHelpers`. The other two are not re-exported from `Concordium.Scheduler.WasmIntegration.V1` and are imported at different types. I think the choice here is reasonable.